src/storage/v2/indices/vector_index.cpp

```
@@ -61,59 +61,124 @@ struct VectorIndex::Impl {
   std::map<std::string, LabelPropKey, std::less<>> index_name_to_label_prop_;
 };
 
+namespace {
+
+/// @brief Attempts to add a vertex to the vector index if it matches the spec criteria.
+/// @param vector_index The vector index to add to.
+/// @param vertex The vertex to potentially add.
+/// @param spec The index specification.
+/// @param snapshot_info Optional snapshot observer for progress tracking.
+void TryAddVertexToIndex(mg_vector_index_t &vector_index, Vertex &vertex, const VectorIndexSpec &spec,
+                         std::optional<SnapshotObserverInfo> const &snapshot_info) {
+  if (!std::ranges::contains(vertex.labels, spec.label_id)) {
+    return;
+  }
+  auto property = vertex.properties.GetProperty(spec.property);
+  if (property.IsNull()) {
+    return;
+  }
+  auto vector = PropertyToFloatVector(property, spec.dimension);
+  if (vector_index.size() >= vector_index.capacity()) {
+    throw query::VectorSearchException(
+        "Vector index is full. Try increasing the capacity.");  // TODO(@DavIvek): Should we do this for user?
+  }
+  vector_index.add(&vertex, vector.data());
```

I think we need to check the result of this `add` and handle errors. Because it can be called concurrently via`PopulateIndexOnMultipleThreads`, we could have two threads concurrently check for capacity, both try to add an item, and for the second one to silently fail.